unit-tests/utilities/concurrency/test-usage.cpp

```diff
@@ -0,0 +1,194 @@
+// License: Apache 2.0. See LICENSE file in root directory.
+// Copyright(c) 2021 Intel Corporation. All Rights Reserved.
+
+//#cmake:add-file ../../../src/dispatcher.cpp
+
+#include "../../test.h"
+#include <common/utilities/time/timer.h>
+#include <src/concurrency.h>
+
+#include <algorithm>
+#include <vector>
+
+using namespace utilities::time;
+
+// We use this function as a CPU stress test function
+int fibo(int num)
+{
+    if (num < 2) return 1;
+    return fibo(num - 1) + fibo(num - 2);
+}
+
+TEST_CASE( "dequeue wait after stop" )
+{
+    single_consumer_queue< std::function< void( void ) > > scq;
+    timer t( std::chrono::seconds( 1 ) );
+    std::function< void( void ) > f;
+    std::function< void( void ) > * f_ptr = &f;
+
+    scq.enqueue( []() {} );
+    REQUIRE( scq.size() == 1 );
+    REQUIRE( scq.peek( &f_ptr ) );
+
+    REQUIRE( scq.started() );
+    REQUIRE_FALSE( scq.stopped() );
+
+    scq.stop();
+
+    REQUIRE_FALSE( scq.peek( &f_ptr ) );
+    REQUIRE( scq.stopped() );
+    REQUIRE_FALSE( scq.started() );
+
+    REQUIRE( scq.empty() );
+
+    t.start();
+    scq.dequeue( &f, 10000 );
+    REQUIRE_FALSE( t.has_expired() );  // Verify no timeout, dequeue return in less than 10 seconds
+}
+
+
+TEST_CASE( "dequeue don't wait when queue is not empty" )
+{
+    single_consumer_queue< std::function< void( void ) > > scq;
+    timer t( std::chrono::seconds( 1 ) );
+    std::function< void( void ) > f;
+
+    scq.enqueue( []() {} );
+    t.start();
+    scq.dequeue( &f, 3000 );
+    REQUIRE_FALSE( t.has_expired() );  // Verify no timeout, dequeue return in less than 1 seconds
+    REQUIRE( scq.empty() );
+}
+
+TEST_CASE( "dequeue wait when queue is empty" )
+{
+    single_consumer_queue< std::function< void( void ) > > scq;
+    timer t( std::chrono::seconds( 1 ) );
+
+    std::function< void( void ) > f;
+
+    t.start();
+    scq.dequeue( &f, 3000 );
+    REQUIRE( t.has_expired() );  // Verify timeout, dequeue return after >= 3 seconds
```

Comment:
Shouldn't `t` be >1 seconds, then?
Wouldn't `dequeue` return `false`, too?

Reply:
I change t to timeout at 2.9 sec and added a dequeue false check